rust/frontend/src/ndarray.rs

```diff
@@ -0,0 +1,359 @@
+    pub fn ctx(&self) -> TVMContext {
+        let dlctx = unsafe { (*self.handle).ctx };
+        dlctx.into()
+    }
+
+    #[inline]
+    pub fn context(&self) -> TVMContext {
+        self.ctx()
+    }
+
+    /// Returns the type of the entries of the NDArray.
+    pub fn dtype(&self) -> TVMType {
+        let dldtype = unsafe { (*self.handle).dtype };
+        dldtype.into()
+    }
+
+    /// Returns the number of dimensions of the NDArray.
+    pub fn ndim(&self) -> usize {
+        unsafe { (*self.handle).ndim as usize }
+    }
+
+    pub fn strides(&self) -> Option<Vec<usize>> {
+        unsafe {
+            (*self.handle).strides.as_mut().map(|pv| {
+                let sz = self.ndim();
+                let mut v: Vec<usize> = Vec::with_capacity(sz * mem::size_of::<usize>());
+                v.as_mut_ptr()
+                    .copy_from_nonoverlapping(pv as *mut _ as *const _, sz);
+                v.set_len(sz);
+                v
+            })
+        }
+    }
+
+    #[inline]
+    pub fn is_contiguous(&self) -> bool {
+        self.strides().is_none()
+    }
+
+    pub fn byte_offset(&self) -> isize {
+        unsafe { (*self.handle).byte_offset as isize }
+    }
+
+    /// Flattens the NDArray to a Vec of the same type in cpu.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// let mut shape = vec![4];
+    /// let mut data = vec![1i32, 2, 3, 4];
+    /// let ctx = TVMContext::cpu(0);
+    /// let mut ndarray = empty(&mut shape, ctx, TVMType::from("int"));
+    /// ndarray.copy_from_buffer(&mut data);
+    /// assert_eq!(ndarray.shape(), Some(shape));
+    /// assert_eq!(ndarray.to_vec::<i32>().unwrap(), data);
+    /// ```
+    pub fn to_vec<T>(&self) -> Result<Vec<T>> {
+        if self.shape().is_none() {
+            return Err(Error::EmptyArray);
+        }
+        let earr = empty(&mut self.shape().unwrap(), TVMContext::cpu(0), self.dtype());
+        let target = self.copy_to_ndarray(earr).unwrap();
+        let arr = unsafe { *(target.handle) };
+        let sz = self.size().unwrap() as usize;
+        let mut v: Vec<T> = Vec::with_capacity(sz * mem::size_of::<T>());
+        unsafe {
+            v.as_mut_ptr()
+                .copy_from_nonoverlapping(arr.data as *const T, sz);
+            v.set_len(sz);
+        }
+        Ok(v)
+    }
+
+    /// Converts the NDArray to [`TVMByteArray`].
+    pub fn to_bytearray(&self) -> Result<TVMByteArray> {
+        let v = self.to_vec::<u8>()?;
+        Ok(TVMByteArray::from(&v))
+    }
+
+    /// Creates an NDArray from a mutable buffer of types i32, u32 or f32 in cpu.
+    /// ## Example
+    ///
+    /// ```
+    /// let mut shape = vec![2];
+    /// let mut data = vec![1f32, 2];
+    /// let ctx = TVMContext::gpu(0);
+    /// let mut ndarray = empty(&mut shape, ctx, TVMType::from("int"));
+    /// ndarray.copy_from_buffer(&mut data);
+    /// ```
+    pub fn copy_from_buffer<T: Num32>(&mut self, data: &mut [T]) {
```

Comment:
please note in the docs that this will panic on the TVM side (can you drop a reference here that shows that this is the case please?)